codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -445,8 +506,9 @@ public FlowControlledBase(final ChannelHandlerContext ctx, final Http2Stream str
             this.endOfStream = endOfStream;
             this.stream = stream;
             this.promise = promise;
-            // Ensure error() gets called in case something goes wrong after the frame is passed to Netty.
-            promise.addListener(this);
+            if (!promise.isVoid()) {
```

Comment:
I'm not against this although there may value in lazy creating a listable promise for all frames types.  We only need to add listeners (and thus create a real promise) if we do write operation.  For example if the frame is cancelled before it gets to write we may not need to bother to create a "real" promise.

Once thing to consider is if anything else happens to the promise (outside the scope of `FlowControlledBase` and children classes), (it is cancelled, set to failure) and it is "able to support listeners" then we must ensure we cleanup.  Can we guarantee that no-one will mess with the promise outside our scope?